File: api/common/src/main/java/net/weavemc/weave/api/Hook.java
package net.weavemc.weave.api;

import org.jetbrains.annotations.NotNull;
import org.objectweb.asm.tree.ClassNode;

public abstract class Hook {

    public final String[] targets;

    public Hook(String... targets) {
        this.targets = targets;
    }

    public abstract void transform(@NotNull ClassNode node, @NotNull AssemblerConfig cfg);

    public interface AssemblerConfig {
        void computeFrames();
    }

}

File: loader/src/main/kotlin/net/weavemc/loader/api/Hook.kt
package net.weavemc.loader.api

import org.objectweb.asm.tree.ClassNode

/**
 * A hook is a type of weave mod intrinsic that allows modders to change the bytecode of any class.
 * Implementors are expected to have a no-args public constructor
 *
 * @property targets the internal names of the targets this [Hook] wants to modify, in the namespace of the mod
 */
public abstract class Hook(public vararg val targets: String) {
    /**
     * Receives all transformed [ClassNode]s in the namespace of the mod that match [targets].
     * The class writer parameters can be altered using the [AssemblerConfig].
     */
    public abstract fun transform(node: ClassNode, cfg: AssemblerConfig)

    /**
     * Allows modders to alter class writer parameters
     */
    public interface AssemblerConfig {
        /**
         * Enables [org.objectweb.asm.ClassWriter.COMPUTE_FRAMES]
         */
        public fun computeFrames()
    }
}
